packages/store/src/keys/keyKind.js

```diff
@@ -0,0 +1,165 @@
+ *    * Otherwise, if it is a valid `copyMap` then we store it in
+ *      the `nonKeyCopyMapMemo` and `return check(false, ...)`.
+ *    * Otherwise, we simply `return check(false,...)`.
+ *
+ * @param {Passable} m
+ * @param {Checker=} check
+ * @returns {boolean}
+ */
+const checkCopyMapIsKey = (m, check = x => x) => {
+  if (nonKeyCopyMapMemo.has(m)) {
+    return check(
+      false,
+      X`This copyMap is not a key because it contains a non-key value: ${m}`,
+    );
+  }
+  // TODO implement
+  return check(false, X`copyMap validation not yet implemented: ${m}`);
+};
+
+/** @type {WeakMap<Key, KeyKind>} */
+const keyKindMemo = new WeakMap();
+
+/**
+ * @param {Passable} val
+ * @param {Checker=} check
+ * @returns {KeyKind=}
+ */
+const checkKeyKindOf = (val, check = x => x) => {
+  if (!isObject(val)) {
+    // @ts-ignore PassStyle and KeyStyle are the same for primitives
+    return passStyleOf(val);
+  }
+  if (keyKindMemo.has(val)) {
+    return keyKindMemo.get(val);
+  }
+  // eslint-disable-next-line no-use-before-define
+  const keyKind = keyKindOfInternal(val, check);
+  if (keyKind !== undefined) {
+    // Don't cache the undefined cases, so that if it is tried again
+    // with `assertChecker` it'll throw a diagnostic again
+    keyKindMemo.set(val, keyKind);
+  }
+  return keyKind;
+};
+
+/**
+ * @param {Passable} val
+ * @param {Checker=} check
+ * @returns {KeyKind=}
+ */
+const keyKindOfInternal = (val, check = x => x) => {
+  const passStyle = passStyleOf(val);
+  switch (passStyle) {
+    case 'copyRecord':
+    case 'copyArray': {
+      const checkIsKey = child => checkKeyKindOf(child, check) !== undefined;
+      if (everyPassableChild(val, checkIsKey)) {
+        // A copyRecord or copyArray is a key iff all its children are keys
+        return passStyle;
+      }
+      return undefined;
+    }
+    case 'tagged': {
+      const tag = getTag(val);
+      switch (tag) {
+        case 'copySet': {
+          if (checkCopySet(val, check)) {
+            return 'copySet';
+          }
+          return undefined;
+        }
+        case 'copyMap': {
+          if (checkCopyMapIsKey(val, check)) {
+            return 'copyMap';
+          }
+          return undefined;
+        }
+        default: {
+          check(false, X`A passable tagged ${q(tag)} is not a key: ${val}`);
+          return undefined;
+        }
+      }
+    }
+    case 'remotable': {
+      // All remotables are keys.
+      return 'remotable';
+    }
+    case 'error':
+    case 'promise': {
+      check(false, X`A ${q(passStyle)} cannot be a key`);
+      return undefined;
+    }
+    default: {
+      // Unexpected tags are just non-keys, but an unexpected passStyle
+      // is always an error.
+      assert.fail(X`unexpected passStyle ${q(passStyle)}: ${val}`);
+    }
+  }
+};
+
+/**
+ * @param {Key} val
+ * @returns {KeyKind}
+ */
+export const keyKindOf = val =>
+  // @ts-ignore When assertChecker is provided, it won't return undefined.
```

Comment:
This ignore is suspicious.  If you don't want to have `undefined` as a potential result, then please `assert(returnValue)` or do something else to strip it out before returning it.

Better to do such an assert for safety rather than just throwing away what TS infers is the type.

Reply:
Done. I wish it wasn't so much more verbose, but it is the right thing to do.